waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,475 @@
+    proc(
+        hashes: seq[WakuMessageHash], peerId: PeerId
+    ): Future[Result[void, string]] {.async: (raises: []), closure.} =
+      var query = StoreQueryRequest()
+      query.includeData = true
+      query.messageHashes = hashes
+      query.paginationLimit = some(uint64(100))
+
+      while true:
+        let catchable = catch:
+          await wakuStoreClient.query(query, peerId)
+
+        if catchable.isErr():
+          return err("store client error: " & catchable.error.msg)
+
+        let res = catchable.get()
+        let response = res.valueOr:
+          return err("store client error: " & $error)
+
+        query.paginationCursor = response.paginationCursor
+
+        for kv in response.messages:
+          let handleRes = catch:
+            await wakuArchive.handleMessage(kv.pubsubTopic.get(), kv.message.get())
+
+          if handleRes.isErr():
+            error "message transfer failed", error = handleRes.error.msg
+            # Messages can be synced next time since they are not added to storage yet.
+            continue
+
+          self.ingessMessage(kv.pubsubTopic.get(), kv.message.get())
+
+        if query.paginationCursor.isNone():
+          break
+
+      return ok()
+  )
+
+proc initFillStorage(
+    self: WakuSync, wakuArchive: WakuArchive
+): Future[Result[void, string]] {.async.} =
+  if wakuArchive.isNil():
+    return err("waku archive unavailable")
+
+  let endTime = getNowInNanosecondTime()
+  let starTime = endTime - self.syncInterval.nanos
+
+  var query = ArchiveQuery(
+    includeData: true,
+    cursor: none(ArchiveCursor),
+    startTime: some(starTime),
+    endTime: some(endTime),
+    pageSize: 100,
+  )
+
+  while true:
+    let response = (await wakuArchive.findMessages(query)).valueOr:
+      return err($error)
+
+    for (topic, msg) in response.topics.zip(response.messages):
+      self.ingessMessage(topic, msg)
+
+    if response.cursor.isNone():
+      break
+
+    query.cursor = response.cursor
+
+  return ok()
+
+proc new*(
+    T: type WakuSync,
+    peerManager: PeerManager,
+    maxFrameSize: int = DefaultMaxFrameSize,
+    syncInterval: timer.Duration = DefaultSyncInterval,
+    relayJitter: Duration = DefaultGossipSubJitter,
+    pruning: bool,
+    wakuArchive: WakuArchive,
+    wakuStoreClient: WakuStoreClient,
+): Future[Result[T, string]] {.async.} =
+  let storage = NegentropyStorage.new().valueOr:
+    return err("negentropy storage creation failed")
+
+  let sync = WakuSync(
+    storage: storage,
+    peerManager: peerManager,
+    maxFrameSize: maxFrameSize,
+    syncInterval: syncInterval,
+    relayJitter: relayJitter,
+    pruneOffset: syncInterval div 100,
+  )
+
+  sync.initProtocolHandler()
+
+  if pruning:
+    sync.initPruningHandler(wakuArchive)
+
+  sync.initTransferHandler(wakuArchive, wakuStoreClient)
+
+  let res = await sync.initFillStorage(wakuArchive)
+  if res.isErr():
+    error "initial storage filling failed", error = res.error
```

Comment:
It will still work but won't be able to sync right away since the storage will not contain any elements to sync.

Maybe warn? It's not a big deal if we sync often.

Reply:
Then it sounds not an error, just would warn the log reader and explain if it will behave different a bit.